src/stan/analyze/mcmc/compute_potential_scale_reduction.hpp

```diff
@@ -16,6 +17,150 @@
 namespace stan {
 namespace analyze {
 
+/**
+ * Computes normalized average ranks for draws. Transforming them to normal
+ * scores using inverse normal transformation and a fractional offset. Based on
+ * paper https://arxiv.org/abs/1903.08008
+ * @param draws stores chains in columns
+ * @return normal scores for average ranks of draws
+ *
+ */
+
+inline Eigen::MatrixXd rank_transform(const Eigen::MatrixXd& draws) {
+  const Eigen::Index rows = draws.rows();
+  const Eigen::Index cols = draws.cols();
+  const Eigen::Index size = rows * cols;
+
+  std::vector<std::pair<double, int>> value_with_index(size);
+
+  for (Eigen::Index i = 0; i < size; ++i) {
+    value_with_index[i] = {draws(i), i};
+  }
+
+  std::sort(value_with_index.begin(), value_with_index.end());
+
+  Eigen::MatrixXd rankMatrix = Eigen::MatrixXd::Zero(rows, cols);
+
+  // Assigning average ranks
+  for (Eigen::Index i = 0; i < size; ++i) {
+    // Handle ties by averaging ranks
+    Eigen::Index j = i + 1;
+    double sumRanks = j;
+    Eigen::Index count = 1;
+
+    while (j < size && value_with_index[j].first == value_with_index[i].first) {
+      sumRanks += j + 1;  // Rank starts from 1
+      ++j;
+      ++count;
+    }
+    double avgRank = sumRanks / count;
+    boost::math::normal_distribution<double> dist;
+    for (std::size_t k = i; k < j; ++k) {
+      Eigen::Index index = value_with_index[k].second;
+      double p = (avgRank - 3.0 / 8.0) / (size - 2.0 * 3.0 / 8.0 + 1.0);
+      rankMatrix(index) = boost::math::quantile(dist, p);
+    }
+    i = j - 1;  // Skip over tied elements
+  }
+  return rankMatrix;
+}
+
+/**
+ * Computes square root of marginal posterior variance of the estimand by the
+ * weigted average of within-chain variance W and between-chain variance B.
+ *
+ * @param draws stores chains in columns
+ * @return square root of ((N-1)/N)W + B/N
+ *
+ */
+
+inline double rhat(const Eigen::MatrixXd& draws) {
+  const Eigen::Index num_chains = draws.cols();
+  const Eigen::Index num_draws = draws.rows();
+
+  Eigen::VectorXd chain_mean(num_chains);
+  chain_mean = draws.colwise().mean();
+  double total_mean = chain_mean.mean();
+  double var_between = num_draws
+                       * (chain_mean.array() - total_mean).square().sum()
+                       / (num_chains - 1);
+  double var_sum = 0;
+  for (Eigen::Index col = 0; col < num_chains; ++col) {
+    var_sum += (draws.col(col).array() - chain_mean(col)).square().sum()
+               / (num_draws - 1);
+  }
+  double var_within = var_sum / num_chains;
```

Comment:
A little more advanced but we can also do this like the below. I added comments to explain the Eigen parts but delete them if you change this.

Also in the below I assume you change `chain_mean` to be Eigen::RowVectorXd`

```suggestion
  double within_variance =
     // Divide each row by chains and get sum of squares for each chain (getting a vector back)
      ((chains.array().rowwise() - chain_means.array())
           .square()
           .colwise()
           .sum() /
       // divide each sum of square by num_draws, sum the sum of squares, and divide by num chains
       (num_draws - 1.0))
          .sum() / num_chains;

```